### PR TITLE
[KOGITO-492] - CLI: Provide option to force operator installation

### PR DESCRIPTION
**Pull request by msmagnanijr**
CLI: Provide option to force operator installation even if present in OperatorHub

See: https://issues.redhat.com/browse/KOGITO-492

Signed-off-by: Mauricio Magnani <mmagnani@redhat.com>

Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.MD#sending-a-pull-request)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [ ] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

### CODE REVIEW

**Comment by ricardozanini** [comment]
@msmagnanijr thanks for your PR! Please, check the requirements.

File: pkg/apis/app/v1alpha1/zz_generated.openapi.go [hunk 1]
```diff
@@ -25,26 +25,26 @@ import (
...
+		"./pkg/apis/app/v1alpha1.KogitoAppStatus":                schema_pkg_apis_app_v1alpha1_KogitoAppStatus(ref),
+		"./pkg/apis/app/v1alpha1.KogitoDataIndex":                schema_pkg_apis_app_v1alpha1_KogitoDataIndex(ref),
+		"./pkg/apis/app/v1alpha1.KogitoDataIndexSpec":            schema_pkg_apis_app_v1alpha1_KogitoDataIndexSpec(ref),
+		"./pkg/apis/app/v1alpha1.KogitoDataIndexStatus":          schema_pkg_apis_app_v1alpha1_KogitoDataIndexStatus(ref),
+		"./pkg/apis/app/v1alpha1.KogitoInfra":                    schema_pkg_apis_app_v1alpha1_KogitoInfra(ref),
+		"./pkg/apis/app/v1alpha1.KogitoInfraSpec":                schema_pkg_apis_app_v1alpha1_KogitoInfraSpec(ref),
+		"./pkg/apis/app/v1alpha1.KogitoInfraStatus":              schema_pkg_apis_app_v1alpha1_KogitoInfraStatus(ref),
+		"./pkg/apis/app/v1alpha1.ResourceMap":                    schema_pkg_apis_app_v1alpha1_ResourceMap(ref),
+		"./pkg/apis/app/v1alpha1.Resources":                      schema_pkg_apis_app_v1alpha1_Resources(ref),
+		"./pkg/apis/app/v1alpha1.WebhookSecret":                  schema_pkg_apis_app_v1alpha1_WebhookSecret(ref),
```
Comment (ricardozanini) [inline pkg/apis/app/v1alpha1/zz_generated.openapi.go]: It seems that you haven't set your Go environment properly: https://golang.org/doc/install

File: cmd/kogito/command/install/kogito_operator.go [hunk 2]
```diff
@@ -76,8 +78,14 @@ func (i *installKogitoOperatorCommand) Exec(cmd *cobra.Command, args []string) e
 		return err
 	}
 
-	if _, err = shared.MustInstallOperatorIfNotExists(i.flags.Namespace, i.flags.Image, i.Client, false); err != nil {
-		return err
+	if _, err = shared.ForceInstallOperator(i.flags.Namespace, i.flags.Image, i.flags.Force, i.Client); err != nil {
+		return nil
+	}
+
+	if !i.flags.Force {
```
Comment (ricardozanini) [inline cmd/kogito/command/install/kogito_operator.go]: Why not refactor to `InstallOperatorIfNotExists` and pass the flag to it? This way we can encapsulate the install algorithm there.

File: cmd/kogito/command/install/kogito_operator.go [hunk 3]
```diff
@@ -23,6 +23,7 @@ import (
 type installKogitoOperatorFlags struct {
 	Namespace string
 	Image     string
+	Force     bool
```
[the remark this answers is not in the document]
Comment (spolti) [reply cmd/kogito/command/install/kogito_operator.go]: If it is configurable it should be a pointer though. otherwise the default value will always False IIRC.

File: cmd/kogito/command/install/kogito_operator.go (hunk 2 above)
Comment (spolti) [reply cmd/kogito/command/install/kogito_operator.go]: +1

File: pkg/apis/app/v1alpha1/zz_generated.openapi.go (hunk 1 above)
Comment (spolti) [reply pkg/apis/app/v1alpha1/zz_generated.openapi.go]: not sure if it is a good idea remove the FQDN repo name and put a dot instead.

File: pkg/apis/app/v1alpha1/zz_generated.openapi.go [hunk 4]
```diff
@@ -1107,7 +1107,7 @@ func schema_pkg_apis_app_v1alpha1_Resources(ref common.ReferenceCallback) common
 			},
 		},
 		Dependencies: []string{
-			"github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1.ResourceMap"},
+			"./pkg/apis/app/v1alpha1.ResourceMap"},
```
Comment (spolti) [inline pkg/apis/app/v1alpha1/zz_generated.openapi.go]: same on all the changes above.

File: cmd/kogito/command/install/kogito_operator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ricardozanini) [reply cmd/kogito/command/install/kogito_operator.go]: doesn't matter in this case since we always have to set a default value to cobra flags.

**Comment by ricardozanini** [comment]
@msmagnanijr also please create a unit test to verify this new feature.

And the tests are failing:
https://app.circleci.com/jobs/github/kiegroup/kogito-cloud-operator/512

Please make sure to run `make test` and `make run-e2e` before sending a PR. Thanks!

File: pkg/apis/app/v1alpha1/zz_generated.openapi.go (hunk 1 above)
Comment (msmagnanijr) [reply pkg/apis/app/v1alpha1/zz_generated.openapi.go]: Sorry, this was automatically generated, I will remove it from the final commit.

Comment (ricardozanini) [reply pkg/apis/app/v1alpha1/zz_generated.openapi.go]: I'm afraid that your Go installation is wrong. The project should be at: `$GOPATH/src/github.com/kiegroup/kogito-cloud-operator`

Comment (msmagnanijr) [reply pkg/apis/app/v1alpha1/zz_generated.openapi.go]: ok! I will check this out.

File: cmd/kogito/command/install/kogito_operator.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (msmagnanijr) [reply cmd/kogito/command/install/kogito_operator.go]: OK! done!

File: cmd/kogito/command/install/kogito_operator.go [hunk 5]
```diff
@@ -76,8 +78,14 @@ func (i *installKogitoOperatorCommand) Exec(cmd *cobra.Command, args []string) e
 		return err
 	}
 
-	if _, err = shared.MustInstallOperatorIfNotExists(i.flags.Namespace, i.flags.Image, i.Client, false); err != nil {
-		return err
+	if _, err = shared.ForceInstallOperator(i.flags.Namespace, i.flags.Image, i.flags.Force, i.Client); err != nil {
```
[the remark this answers is not in the document]
Comment (msmagnanijr) [reply cmd/kogito/command/install/kogito_operator.go]: done!

File: cmd/kogito/command/install/kogito_operator.go (hunk 2 above)
Comment (msmagnanijr) [reply cmd/kogito/command/install/kogito_operator.go]: @ricardozanini do you say "MustInstallOperatorIfNotExists"? It is used in many other places in the code. This change would impact several places in the project, which is unnecessary as the only place that needs to "force" is the installation. Makes sense? thanks

Comment (radtriste) [reply cmd/kogito/command/install/kogito_operator.go]: I don't think we need a specific behavior and specific function. 
The `force` is specific to the CLI execution. I would just keep it in the exec command if possible:
```
if i.flags.Force {
    // InstallOperatorWithYamlFiles <= We would need to make it public
} else {
    // MustInstallOperatorIfNotExists
}
```
But could also make sense to update `MustInstallOperatorIfNotExists` now with a new boolean if we plan to use this `force` in another place ?

File: cmd/kogito/command/shared/resource_checks.go [hunk 6]
```diff
@@ -83,7 +83,8 @@ func checkKogitoOperatorExists(kubeCli *client.Client, namespace string) (bool,
 	}
 
 	if operatorDeployment.Status.AvailableReplicas == 0 {
-		return true, fmt.Errorf("Kogito Operator seems to be created in the namespace '%s', but there's no available pods replicas deployed ", namespace)
+		log.Warnf("Kogito Operator seems to be created in the namespace '%s' but there's no available pods replicas deployed ", namespace)
```
Comment (radtriste) [inline cmd/kogito/command/shared/resource_checks.go]: I did not get why this was changed to a warn ?
There should be an error if no available replica or ?

File: cmd/kogito/command/install/kogito_operator.go (hunk 2 above)
Comment (ricardozanini) [reply cmd/kogito/command/install/kogito_operator.go]: Please don’t expose the Install...YamlFiles. This function is intended to be private and encapsulate a way to install the operator. The delegate function (Must..) is responsible to call it. I will wrote another function to deal with subscriptions later that will make use of it. 

The purpose is to create another delegate function that just installs the operator whether or not it’s installed in the cluster. 

And @msmagnanijr please remove the alias “f” of the flag since I will use it to “follow” output logs in future releases.

File: cmd/kogito/command/shared/resource_checks.go [hunk 7]
```diff
@@ -83,7 +83,8 @@ func checkKogitoOperatorExists(kubeCli *client.Client, namespace string) (bool,
...
+		return true, nil
```
Comment (radtriste) [inline cmd/kogito/command/shared/resource_checks.go]: Why did you change this ?
How can we distinguish an operator existing and running to one which is not existing ?
Or should we provide a new method `checkKogitoOperatorRunning` that checks the number of available replica as well (and we could remove this part) ?

**Comment by ricardozanini** [comment]
Solved here: https://github.com/kiegroup/kogito-cloud-operator/pull/176
